src/cluster/serializer/broker/Serializer.cc

```diff
@@ -0,0 +1,219 @@
+#include "zeek/cluster/serializer/broker/Serializer.h"
+
+#include <optional>
+
+#include "zeek/Desc.h"
+#include "zeek/Func.h"
+#include "zeek/Reporter.h"
+#include "zeek/broker/Data.h"
+#include "zeek/cluster/Backend.h"
+
+#include "broker/data_envelope.hh"
+#include "broker/error.hh"
+#include "broker/format/json.hh"
+#include "broker/zeek.hh"
+
+#include "zeek/3rdparty/doctest.h"
+
+using namespace zeek::cluster;
+
+namespace {
+
+/**
+ * Convert a cluster::detail::Event to a broker::zeek::Event.
+ *
+ * @param ev The cluster::detail::Event
+ * @return A broker::zeek::Event to be serialized, or nullopt in case of errors.
+ */
+std::optional<broker::zeek::Event> to_broker_event(const detail::Event& ev) {
+    broker::vector xs;
+    xs.reserve(ev.args.size());
+    for ( const auto& a : ev.args ) {
+        if ( auto res = zeek::Broker::detail::val_to_data(a.get()) ) {
+            xs.emplace_back(std::move(res.value()));
+        }
+        else {
+            return std::nullopt;
+        }
+    }
+
+    return broker::zeek::Event(ev.HandlerName(), xs, broker::to_timestamp(ev.timestamp));
+}
+
+/**
+ * Convert a broker::zeek::Event to cluster::detail::Event by looking
+ * it up in Zeek's event handler registry and converting event arguments
+ * to the appropriate Val instances.
+ *
+ * @param broker_ev The broker side event.
+ * @returns A zeek::cluster::detail::Event instance, or std::nullopt if the conversion failed.
+ */
+std::optional<detail::Event> to_zeek_event(const broker::zeek::Event& ev) {
+    auto&& name = ev.name();
+    auto&& args = ev.args();
+
+    // Meh, technically need to convert ev.metadata() and
+    // expose it to script land as `table[count] of any`
+    // where consumers then know what to do with it.
+    //
+    // For now, handle the timestamp explicitly.
+    double ts;
+    if ( auto ev_ts = ev.ts() )
+        broker::convert(*ev_ts, ts);
+    else
+        ts = zeek::run_state::network_time;
+
+    zeek::Args vl;
+    zeek::EventHandlerPtr handler = zeek::event_registry->Lookup(name);
+    if ( ! handler ) {
+        zeek::reporter->Error("Failed to lookup handler for '%s'", std::string(name).c_str());
```

Comment:
Yeah, that's interesting. I'm not doing something here now, but maybe something akin to passing the plugin instance to debug streams?

If one assumes all code is in form of plugins, this might work out okay:
```
zeek::reporter->Info(plugin, ...)
zeek::reporter->Error(plugin, ...)
```

There's also the thread-safety concern, maybe we could make #3998 more generic, too.